csrc/device_lower/pass/index.cpp

```diff
@@ -1816,6 +1821,224 @@ Val* hardCodedIndexGenerationForStMatrix(
+
+  NVF_ERROR(
+      ldst->out()->dtype() == DataType::Half,
+      "we only support half type in stmatrix");
+
+  NVF_ERROR(ldst->out()->isA<TensorView>());
+  TensorView* out_tv = ldst->out()->as<TensorView>();
+  NVF_ERROR(getSwizzle(out_tv) == MmaInputSmemSwizzle::B128);
+
+  // Constants
+  constexpr int64_t dtype_size = 2;
+  constexpr int64_t warp_size = 32;
+  constexpr int64_t swizzle_row_size = 8;
+  constexpr int64_t stsm_column_size = 8;
+  constexpr int64_t swizzle_n_tile = 64;
+  constexpr int64_t swizzle_n_iter = 4;
+
+  // Derived constants
+  constexpr int64_t stsm_column_stride = stsm_column_size * dtype_size;
+  const int64_t swizzle_n_tile_stride = swizzle_n_tile * dtype_size;
+  const int64_t stsm_n_tile_stride = stsm_n_tile / stsm_column_size;
+  const int64_t tile_stride = tma_m * swizzle_n_tile * dtype_size;
+  const int64_t tdy_stride = tma_m * tma_n * dtype_size;
+
+  // NvFuser Val for constants
+  Val* warp_size_val = IrBuilder::create<Val>(warp_size, DataType::Index);
+  Val* stsm_m_tile_val = IrBuilder::create<Val>(stsm_m_tile, DataType::Index);
+  Val* stsm_n_tile_val = IrBuilder::create<Val>(stsm_n_tile, DataType::Index);
+  Val* stsm_n_tile_stride_val =
+      IrBuilder::create<Val>(stsm_n_tile_stride, DataType::Index);
+  Val* swizzle_row_size_val =
+      IrBuilder::create<Val>(swizzle_row_size, DataType::Index);
+  Val* stsm_column_stride_val =
+      IrBuilder::create<Val>(stsm_column_stride, DataType::Index);
+  Val* swizzle_n_tile_stride_val =
+      IrBuilder::create<Val>(swizzle_n_tile_stride, DataType::Index);
+  Val* tile_stride_val = IrBuilder::create<Val>(tile_stride, DataType::Index);
+  Val* tdy_stride_val = IrBuilder::create<Val>(tdy_stride, DataType::Index);
+  Val* swizzle_n_iter_val =
+      IrBuilder::create<Val>(swizzle_n_iter, DataType::Index);
+
+  // Derived Constants
+  NamedScalar* TDX =
+      IrBuilder::create<NamedScalar>("threadIdx.x", DataType::Index);
+  NamedScalar* TDY =
+      IrBuilder::create<NamedScalar>("threadIdx.y", DataType::Index);
+  Val* warp_id = SimplifyingIrBuilder::divExpr(TDX, warp_size_val);
+  Val* lane_id = SimplifyingIrBuilder::modExpr(TDX, warp_size_val);
+
+  Val* outer_index =
+      SimplifyingIrBuilder::divExpr(loop->index(), swizzle_n_iter_val);
+  Val* inner_index =
+      SimplifyingIrBuilder::modExpr(loop->index(), swizzle_n_iter_val);
+
+  // Calculate Row
+  Val* warp_row = SimplifyingIrBuilder::mulExpr(warp_id, stsm_m_tile_val);
+  Val* lane_row = SimplifyingIrBuilder::modExpr(lane_id, stsm_m_tile_val);
+  Val* row = SimplifyingIrBuilder::addExpr(warp_row, lane_row);
+
+  // Calculate Column
+  Val* lane_col = SimplifyingIrBuilder::divExpr(lane_id, stsm_n_tile_val);
+  Val* iter_col =
+      SimplifyingIrBuilder::mulExpr(inner_index, stsm_n_tile_stride_val);
+  Val* col = SimplifyingIrBuilder::addExpr(lane_col, iter_col);
+
+  // Swizzle Column
+  Val* row_in_swizzle_pattern =
+      SimplifyingIrBuilder::modExpr(row, swizzle_row_size_val);
```

Comment:
Is there a way to use expression simplifier or scalar hoisting to reuse `row` value instead of recomputing it twice?

Reply:
I believe you can do:
```C++
out_index = GpuLower::current()->commonScalarMap().hoistScalar(out_index, for_loops_);
```
in the end of this function, and it will automatically be hoisted.